### PR TITLE
Update documentation to reflect new catalog features

### PR DESCRIPTION
**Pull request by dramaticlly**
I want to include new features we developed over last few releases to be reflected in javadoc of SparkCatalog as it's lacking at the moment

- catalog default/override in #4011
- rest catalog in #5570 
- catalog io-impl in #1618 
- catalog metrics-reporter-impl in #6404

CC @jackye1995 @Fokko @SinghAsDev @nastra

### CODE REVIEW

**Comment by jackye1995** [comment]
Could you also do it for Spark 3.4?

**Comment by jackye1995** [comment]
oh nvm about last comment, clicked the wrong link only showing the first commit.

File: docs/spark-configuration.md [hunk 1]
```diff
@@ -63,15 +63,20 @@ Iceberg supplies two implementations:
...
+| spark.sql.catalog._catalog-name_.metrics-reporter-impl        |                               | The custom MetricsReporter implementation.                                                                                                                                                                                               |
+| spark.sql.catalog._catalog-name_.default-namespace            | default                       | The default current namespace for the catalog                                                                                                                                                                                            |
+| spark.sql.catalog._catalog-name_.uri                          | thrift://host:port            | Hive Metastore URI (default from `hive-site.xml`) for hive catalog and REST URI for rest catalog                                                                                                                                         |
```
Comment (jackye1995) [inline docs/spark-configuration.md:73]: should also add example for a REST URI

Comment (dramaticlly) [reply docs/spark-configuration.md:73]: Sure I can add one such example for rest catalog like hive and Hadoop catalog above

File: docs/spark-configuration.md [hunk 2]
```diff
@@ -63,15 +63,20 @@ Iceberg supplies two implementations:
...
-| spark.sql.catalog._catalog-name_.uri               | thrift://host:port            | Metastore connect URI; default from `hive-site.xml` |
-| spark.sql.catalog._catalog-name_.warehouse         | hdfs://nn:8020/warehouse/path | Base path for the warehouse directory |
-| spark.sql.catalog._catalog-name_.cache-enabled     | `true` or `false`             | Whether to enable catalog cache, default value is `true` |
-| spark.sql.catalog._catalog-name_.cache.expiration-interval-ms | `30000` (30 seconds) | Duration after which cached catalog entries are expired; Only effective if `cache-enabled` is `true`. `-1` disables cache expiration and `0` disables caching entirely, irrespective of `cache-enabled`. Default is `30000` (30 seconds) |                                                   |
+| Property                                                      | Values                        | Description                                                                                                                                                                                                                              |
```
Comment (jackye1995) [inline docs/spark-configuration.md:66]: should we add a column for the default?

File: docs/spark-configuration.md [hunk 3]
```diff
@@ -63,15 +63,20 @@ Iceberg supplies two implementations:
...
+| spark.sql.catalog._catalog-name_.cache.expiration-interval-ms | `30000` (30 seconds)          | Duration after which cached catalog entries are expired; Only effective if `cache-enabled` is `true`. `-1` disables cache expiration and `0` disables caching entirely, irrespective of `cache-enabled`. Default is `30000` (30 seconds) |                                                   |
+| spark.sql.catalog._catalog-name_.table-default.propertyKey    |                               | Iceberg table property $propertyKey default at catalog level. A different table property value can be overridden by user                                                                                                                 |
+| spark.sql.catalog._catalog-name_.table-override.propertyKey   |                               | Iceberg table property $propertyKey enforced at catalog level. Cannot be overridden by user                                                                                                                                              |
```
Comment (jackye1995) [inline docs/spark-configuration.md:78]: nit: `_property-key_`, following same pattern as `_catalog-name_`

**Comment by jackye1995** [comment]
Looks like CI failed, could you check that?

File: docs/spark-configuration.md (hunk 1 above)
Comment (nastra) [inline docs/spark-configuration.md:73]:
```suggestion
| spark.sql.catalog._catalog-name_.uri                          | thrift://host:port            | Hive Metastore URI (default from `hive-site.xml`) for hive catalog and REST URI for REST catalog                                                                                                                                         |
```

File: docs/spark-configuration.md [hunk 4]
```diff
@@ -63,15 +63,20 @@ Iceberg supplies two implementations:
...
+| Property                                                      | Values                        | Description                                                                                                                                                                                                                              |
+|---------------------------------------------------------------|-------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| spark.sql.catalog._catalog-name_.type                         | `hive`, `hadoop` or `rest`    | The underlying Iceberg catalog implementation, `HiveCatalog`, `HadoopCatalog`, `RESTCatalog` or left unset if using a custom catalog                                                                                                     |
```
[the remark this answers is not in the document]
Comment (dramaticlly) [reply docs/spark-configuration.md:68]: thank you @nastra . I think in intellij IDE it's easy correction to realign the misformatted mark down table thats why I just apply the realignment. 

I do see it can cause additional difficulty for the review, but I want to share that the github review have a rich text diff for mark down syntax so it can show how it looks like when reading the rendered doc directly like in 

image

let me know if you hate it, I can always revert back to original format

File: docs/spark-configuration.md (hunk 2 above)
Comment (dramaticlly) [reply docs/spark-configuration.md:66]: good point @jackye1995 , I was thinking about the same but looking at current catalog properties listed above, I dont think we actually default much other than (cache-enabled and expiration-interval-ms) per https://github.com/apache/iceberg/blob/master/core/src/main/java/org/apache/iceberg/CatalogProperties.java. 

do you have some default want to highlight into the table above? please let me know and I can add it

Comment (jackye1995) [reply docs/spark-configuration.md:74]: I see, in that case it's probably sufficient to say those defaults in descriptions. Also worth mentioning that if `type` is null, `catalog-impl` must not be null.

File: docs/spark-configuration.md [hunk 5]
```diff
@@ -40,6 +40,14 @@ spark.sql.catalog.hive_prod.uri = thrift://metastore-host:port
 # omit uri to use the same URI as Spark: hive.metastore.uris in hive-site.xml
 ```
 
+Below is example for rest catalog named `rest_prod` that loads tables from REST URL `http://localhost:8080`:
```
Comment (nastra) [inline docs/spark-configuration.md:43]:
```suggestion
Below is an example for a REST catalog named `rest_prod` that loads tables from REST URL `http://localhost:8080`:
```

File: docs/spark-configuration.md [hunk 6]
```diff
@@ -63,15 +71,20 @@ Iceberg supplies two implementations:
...
+| spark.sql.catalog._catalog-name_.warehouse                    | hdfs://nn:8020/warehouse/path | Base path for the warehouse directory                                                                                                                                                                                                    |
+| spark.sql.catalog._catalog-name_.cache-enabled                | `true` or `false`             | Whether to enable catalog cache, default value is `true`                                                                                                                                                                                 |
+| spark.sql.catalog._catalog-name_.cache.expiration-interval-ms | `30000` (30 seconds)          | Duration after which cached catalog entries are expired; Only effective if `cache-enabled` is `true`. `-1` disables cache expiration and `0` disables caching entirely, irrespective of `cache-enabled`. Default is `30000` (30 seconds) |                                                   |
```
Comment (nastra) [inline docs/spark-configuration.md:84]:
```suggestion
| spark.sql.catalog._catalog-name_.cache.expiration-interval-ms | `30000` (30 seconds)          | Duration after which cached catalog entries are expired; Only effective if `cache-enabled` is `true`. `-1` disables cache expiration and `0` disables caching entirely, irrespective of `cache-enabled`. Default is `30000` (30 seconds) |
```

File: docs/spark-configuration.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (dramaticlly) [reply docs/spark-configuration.md]: cool @nastra , I reverted and now diff exclude formatting changes

File: docs/spark-configuration.md [hunk 7]
```diff
@@ -66,12 +74,16 @@ Both catalogs are configured using properties nested under the catalog name. Com
 | Property                                           | Values                        | Description                                                          |
 | -------------------------------------------------- | ----------------------------- | -------------------------------------------------------------------- |
 | spark.sql.catalog._catalog-name_.type              | `hive`, `hadoop` or `rest`    | The underlying Iceberg catalog implementation, `HiveCatalog`, `HadoopCatalog`, `RESTCatalog` or left unset if using a custom catalog |
-| spark.sql.catalog._catalog-name_.catalog-impl      |                               | The underlying Iceberg catalog implementation.|
+| spark.sql.catalog._catalog-name_.catalog-impl      |                               | The custom Iceberg catalog implementation. if `type` is null, `catalog-impl` must not be null. |
```
Comment (nastra) [inline docs/spark-configuration.md:77]:
```suggestion
| spark.sql.catalog._catalog-name_.catalog-impl      |                               | The custom Iceberg catalog implementation. If `type` is null, `catalog-impl` must not be null. |
```

File: docs/spark-configuration.md [hunk 8]
```diff
@@ -66,12 +74,16 @@ Both catalogs are configured using properties nested under the catalog name. Com
...
+| spark.sql.catalog._catalog-name_.io-impl                      |                               | The custom FileIO implementation.                                                                                                                                                                                                        |
```
Comment (nastra) [inline docs/spark-configuration.md:78]:
```suggestion
| spark.sql.catalog._catalog-name_.io-impl                      |                               | The custom FileIO implementation. |
```

File: docs/spark-configuration.md [hunk 9]
```diff
@@ -66,12 +74,16 @@ Both catalogs are configured using properties nested under the catalog name. Com
...
+| spark.sql.catalog._catalog-name_.metrics-reporter-impl        |                               | The custom MetricsReporter implementation.                                                                                                                                                                                               |
```
Comment (nastra) [inline docs/spark-configuration.md:79]:
```suggestion
| spark.sql.catalog._catalog-name_.metrics-reporter-impl        |                               | The custom MetricsReporter implementation.  |
```

File: docs/spark-configuration.md [hunk 10]
```diff
@@ -66,12 +74,16 @@ Both catalogs are configured using properties nested under the catalog name. Com
...
 | spark.sql.catalog._catalog-name_.cache-enabled     | `true` or `false`             | Whether to enable catalog cache, default value is `true` |
-| spark.sql.catalog._catalog-name_.cache.expiration-interval-ms | `30000` (30 seconds) | Duration after which cached catalog entries are expired; Only effective if `cache-enabled` is `true`. `-1` disables cache expiration and `0` disables caching entirely, irrespective of `cache-enabled`. Default is `30000` (30 seconds) |                                                   |
+| spark.sql.catalog._catalog-name_.cache.expiration-interval-ms | `30000` (30 seconds) | Duration after which cached catalog entries are expired; Only effective if `cache-enabled` is `true`. `-1` disables cache expiration and `0` disables caching entirely, irrespective of `cache-enabled`. Default is `30000` (30 seconds) |
+| spark.sql.catalog._catalog-name_.table-default._propertyKey_  |                               | Iceberg table property _propertyKey_ default at catalog level. A different table property value can be overridden by user                                                                                                                |
```
Comment (nastra) [inline docs/spark-configuration.md:85]:
```suggestion
| spark.sql.catalog._catalog-name_.table-default._propertyKey_  |                               | Iceberg table property _propertyKey_ default at catalog level. A different table property value can be overridden by the user |
```

File: docs/spark-configuration.md [hunk 11]
```diff
@@ -66,12 +74,16 @@ Both catalogs are configured using properties nested under the catalog name. Com
...
+| spark.sql.catalog._catalog-name_.table-override._propertyKey_ |                               | Iceberg table property _propertyKey_ enforced at catalog level. Cannot be overridden by user                                                                                                                                             |
```
Comment (nastra) [inline docs/spark-configuration.md:86]:
```suggestion
| spark.sql.catalog._catalog-name_.table-override._propertyKey_ |                               | Iceberg table property _propertyKey_ enforced at catalog level. Cannot be overridden by the user |
```

**Review by nastra — APPROVED** [review]
there were still a few unnecessary whitespace changes in the Table. Once those are fixed the changes LGTM

File: docs/spark-configuration.md [hunk 12]
```diff
@@ -66,12 +74,16 @@ Both catalogs are configured using properties nested under the catalog name. Com
...
 | spark.sql.catalog._catalog-name_.cache-enabled     | `true` or `false`             | Whether to enable catalog cache, default value is `true` |
-| spark.sql.catalog._catalog-name_.cache.expiration-interval-ms | `30000` (30 seconds) | Duration after which cached catalog entries are expired; Only effective if `cache-enabled` is `true`. `-1` disables cache expiration and `0` disables caching entirely, irrespective of `cache-enabled`. Default is `30000` (30 seconds) |                                                   |
+| spark.sql.catalog._catalog-name_.cache.expiration-interval-ms | `30000` (30 seconds) | Duration after which cached catalog entries are expired; Only effective if `cache-enabled` is `true`. `-1` disables cache expiration and `0` disables caching entirely, irrespective of `cache-enabled`. Default is `30000` (30 seconds) |
+| spark.sql.catalog._catalog-name_.table-default._propertyKey_  |                               | Iceberg table property _propertyKey_ default at catalog level. A different table property value can be overridden by the user |
```
Comment (szehon-ho) [inline docs/spark-configuration.md:85]: I think the first sentence can be clarified a little bit: what do you think?


Default Iceberg table property value for property key `_propertyKey_,` which will be set on tables created by this catalog.  These properties can be overridden in the createTable call for individual tables

Hope its not too wordy

File: docs/spark-configuration.md [hunk 13]
```diff
@@ -66,12 +74,16 @@ Both catalogs are configured using properties nested under the catalog name. Com
...
+| spark.sql.catalog._catalog-name_.table-override._propertyKey_ |                               | Iceberg table property _propertyKey_ enforced at catalog level. Cannot be overridden by the user |
```
Comment (szehon-ho) [inline docs/spark-configuration.md:86]: Similar vein:

Default Iceberg table property value for property key `_property_key`, which will be set on tables created by this catalog.  These properties cannot be overridden.

**Review by szehon-ho — COMMENTED** [review]
Looks great, just a few suggestions.  Hope its not repeating any of the above comments

File: docs/spark-configuration.md (hunk 12 above)
Comment (dramaticlly) [reply docs/spark-configuration.md:85]: thank you for the suggestion @szehon-ho . I think these default properties can also be override on updateTable call, do you think we shall add as well or it might be too descriptive on how?

File: docs/spark-configuration.md (hunk 13 above)
Comment (dramaticlly) [reply docs/spark-configuration.md:86]: I am a bit worried that we might want to highlight the difference between `table-default` and `table-override `, so maybe should not use `Default` here? what do you think?

Comment (szehon-ho) [reply docs/spark-configuration.md:86]: Yea good point !

File: docs/spark-configuration.md (hunk 12 above)
Comment (szehon-ho) [reply docs/spark-configuration.md:85]: Ah ok .. hm how about?

Default Iceberg table property value for property key _propertyKey_, which will be set on tables created by this catalog if not overridden.

Comment (dramaticlly) [reply docs/spark-configuration.md]: thank you @szehon-ho , updated per your suggestion.

**Review by szehon-ho — APPROVED** [review]

**Comment by szehon-ho** [comment]
Merged, thanks @dramaticlly , @nastra @jackye1995 for review
